### DISCUSSION TITLE
Recommended way to know when `Torrent.DownloadPieces` is done

### DISCUSSION THREAD

**Discussion by 0x-r4bbit** [anacrolix/torrent#710] [Q&A] [answered]
Hi everyone,

I'm trying to write program where I selectively decide which pieces of a given torrent should be downloaded. `DownloadPieces` seems to be the way to go here as it enables me to do exactly that.

However, it seems that, what it really does, is marking the given pieces for download, which then implicitly triggers the download.

I'm wondering, what's the best way to get a handle of the download in progress? Assuming I call `torrent.DownloadPieces(3, 6)` how can I get informed that the download is done?

@anacrolix thanks in advance and also massive thanks for creating/maintaining this library!

**Comment by anacrolix** [comment, marked as answer]
I think there's a very long existing issue on this, I'll link you later.

**Reply by anacrolix** [reply]
I thought there was a more comprehensive issue but I didn't find it: https://github.com/anacrolix/torrent/issues/168. Feel free to discuss here.

**Reply by 0x-r4bbit** [reply]
Thanks for linking that issue! It definitely has some useful information in there.
What I did now was also subscribing to piece state changes, but since I need to know whether particular pieces are done downloading, I'm checking for these pieces `PieceState.Completed` value.

Something along those lines:

```go
psc := torrent.SubscribePieceStateChanges()
<- psc.Values
if torrent.PieceState(6).Complete && torrent.PieceState(7).Complete && torrent.PieceState(8).Complete {
    log.Print("Download complete!!")
    psc.Close()
}
```

Does this make sense?

**Reply by anacrolix** [reply]
Yes that sounds right.
